File: retrocpu_test_framework/src/hex_cdb.ts
/**
 * Intel HEX / SDCC CDB の生成
 * 根拠: test_framework.mdc / emulater_code_test.mdc
 */

/**
 * 1 レコードのチェックサム（2 の補数）。
 * @param bytes 長さ・アドレス・種別・データを含むレコード本体
 * @returns 0–255
 */
function checksum(bytes: number[]): number {
  let sum = 0;
  for (const b of bytes) sum = (sum + b) & 0xff;
  return (~sum + 1) & 0xff;
}

/**
 * リンク済みバイトイメージを Intel HEX テキストにする。
 * 全ゼロの 16 バイト塊は省略する（ロード先は事前ゼロクリア前提）。
 * @param image ビッグエンディアン・先頭＝バイト 0
 * @returns IHX 全文（EOF 付き）
 */
export function imageToIntelHex(image: Uint8Array): string {
  const lines: string[] = [];
  for (let addr = 0; addr < image.length; addr += 16) {
    const chunk = image.subarray(addr, Math.min(addr + 16, image.length));
    if (chunk.every((b) => b === 0)) {
      continue;
    }
    const rec = [
      chunk.length,
      (addr >> 8) & 0xff,
      addr & 0xff,
      0,
      ...chunk,
    ];
    rec.push(checksum(rec));
    lines.push(
      `:${rec.map((b) => b.toString(16).padStart(2, "0").toUpperCase()).join("")}`,
    );
  }
  lines.push(":00000001FF");
  return `${lines.join("\n")}\n`;
}

/**
 * sdld マップ（または同等の名前→バイトアドレス表）を SDCC CDB の L:G にする。
 * 製品リンクは sdld 本体と `defsToCdbFromSdld`（retrocpu_asm）が担う。
 * ここはマップフォールバック／単体テスト用。`__CP$` はチェックポイント側へ。
 * @param defs シンボル名 → バイトアドレス
 * @returns CDB テキスト
 */
export function defsToCdb(defs: Map<string, number>): string {
  const names = [...defs.keys()]
    .filter((n) => !/^__CP[0-9]{4}$/i.test(n) && !/^__CP\$/i.test(n))
    .sort((a, b) => a.localeCompare(b));
  const lines: string[] = [];
  for (const name of names) {
    const byteAddr = defs.get(name)! >>> 0;
    const hex = byteAddr.toString(16).toUpperCase();
    lines.push(`L:G$${name}$0$0:${hex}`);
  }
  return `${lines.join("\n")}\n`;
}

File: retrocpu_test_framework_rs/src/hex_cdb.rs
use std::collections::HashMap;

fn checksum(bytes: &[u8]) -> u8 {
    let sum = bytes.iter().fold(0_u16, |acc, &b| (acc + b as u16) & 0xff);
    (!sum as u8).wrapping_add(1)
}

pub fn image_to_intel_hex(image: &[u8]) -> String {
    let mut lines: Vec<String> = Vec::new();

    let mut addr: usize = 0;
    while addr < image.len() {
        let end = (addr + 16).min(image.len());
        let chunk = &image[addr..end];

        if chunk.iter().all(|&b| b == 0) {
            addr += 16;
            continue;
        }

        let mut rec: Vec<u8> = Vec::with_capacity(5 + chunk.len());
        rec.push(chunk.len() as u8);
        rec.push(((addr >> 8) & 0xff) as u8);
        rec.push((addr & 0xff) as u8);
        rec.push(0);
        rec.extend_from_slice(chunk);
        rec.push(checksum(&rec));

        let line = format!(
            ":{}",
            rec.iter()
                .map(|b| format!("{:02X}", b))
                .collect::<Vec<_>>()
                .join("")
        );
        lines.push(line);

        addr += 16;
    }

    lines.push(":00000001FF".to_string());
    format!("{}\n", lines.join("\n"))
}

pub fn defs_to_cdb(defs: &HashMap<String, u32>) -> String {
    let mut names: Vec<&String> = defs
        .keys()
        .filter(|n| {
            let up = n.to_ascii_uppercase();
            !(up.starts_with("__CP$")
                || (up.len() == 8
                    && up.starts_with("__CP")
                    && up[4..].chars().all(|c| c.is_ascii_digit())))
        })
        .collect();
    names.sort();

    let mut lines = Vec::with_capacity(names.len());
    for name in names {
        let byte_addr = defs.get(name).copied().unwrap_or(0);
        lines.push(format!("L:G${}$0$0:{:X}", name, byte_addr));
    }

    format!("{}\n", lines.join("\n"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn intel_hex_skips_zero_chunks_and_has_eof() {
        let mut image = vec![0_u8; 64];
        image[0x10] = 0x12;
        image[0x11] = 0x34;
        let hex = image_to_intel_hex(&image);
        assert!(hex.contains(":100010001234"));
        assert!(hex.ends_with(":00000001FF\n"));
    }

    #[test]
    fn cdb_filters_checkpoint_symbols() {
        let mut defs = HashMap::new();
        defs.insert("gl_main".to_string(), 0x200);
        defs.insert("__CP0001".to_string(), 0x1234);
        defs.insert("__CP$foo$0001".to_string(), 0x1234);
        let cdb = defs_to_cdb(&defs);
        assert!(cdb.contains("L:G$gl_main$0$0:200"));
        assert!(!cdb.contains("__CP0001"));
        assert!(!cdb.contains("__CP$foo$0001"));
    }
}
